src/mongo_client.ts

```diff
@@ -409,44 +411,59 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
       return this;
     }
 
-    const options = this[kOptions];
+    if (this.connectionLock) {
```

Comment:
Hey @clemclx - thanks for the fix!  I think we're on the right track here.

We can't use them (yet) but I think a decorator might be a good fit here.  

```typescript
class MongoClient {
  ...
  @lock 
  connect() { 
    // no changes to method body here
  }
}
```

What do you think of a slight refactor here, to make the refactor to use a generator (when it's available) straightforward?  I was thinking we could

- move the entire (old) body of `connect` into a private `_connect` method
- make the current `connect` method enforce locking on the `_connect` method

Then, porting the logic to a `@lock` generator is straightforward because we just move the code from `_connect` back into `connect` and move the locking logic into the generator.  Thoughts?

---

as far as testing, I think Anna is correct.  It should be enough to parallelize a few connect invocations and assert that

- we successfully connect
- we only receive one `Topology.OPEN` event

These tests can live in `integration/mongo_client.test.ts`.

lmk if you want me to add the tests for you but feel free to add them if you'd like to!

Reply:
Hey @baileympearson, thanks for your review !
I agree that a decorator would fit pretty well here, make sense. Did not took this way as I did not see any decorator in the driver codebase (you precises that it's not supported yet anyway).

I refactored the code to fit your needs, let me know if I correctly understood your request.

Moreover, I added the test suite as you suggested but I think there is something wrong with the leak_checker hook (`tools/runner/hooks/leak_checker.ts`)...
The `afterEach` hook fails on the first test as it seems to count the number of `connect` called on the prototype (instead maybe of couting the effective number of connections opens ?). The problem is that calling `client.connect` thrice to make concurrent calls increase the counter to 3 but calling `client.close` once make the assertion fails as it only decrease one.
A workaround could be to call the `client.close` as many times as `client.connect` but I really don't think it's the goal here.
I don't have the big picture of this hook to properly understand its purpose and constraints, could you check it please ?

Let me know if I can help out !

Reply:
Hey @clemclx - yup, this is basically what I had in mind.

That's an interesting point about the leak checker.  Historically, we've leaked connections by not closing the client so the leak checker was added to ensure that we close every client we open in our tests.  I think it's fine to close it as many times as we need to for these tests (and add a comment explaining why), since this is a unique test where we explicitly connect a single client multiple times.

Reply:
Hello @baileympearson ! Thanks for the clarification about the leak_checker.
I applied the fix to tests in consequences.
We should be fine now, let me know if I can help for something else :smiley: